File: netlify/lib/parser.ts
/** Extração limitada e somente leitura do XLSX: apenas valores em cache, com origem da célula.
    Porte fiel de backend/app/parser.py — as regras de recusa e os avisos são os mesmos. */

import { unzipSync } from "fflate";
import * as XLSX from "xlsx";
import { CATALOG, type Indicator } from "./catalog.js";

export const MAX_ROWS = 30000;
export const MAX_COLS = 64;
const MONTHS = [
  "janeiro",
  "fevereiro",
  "marco",
  "abril",
  "maio",
  "junho",
  "julho",
  "agosto",
  "setembro",
  "outubro",
  "novembro",
  "dezembro",
];

export class WorkbookError extends Error {}

export interface RecordValue {
  year: number;
  month: number | null;
  house: string | null;
  location: string | null;
  category: string | null;
  value: number | null;
  cell: string;
}

export interface ParsedIndicator {
  code: string;
  title: string;
  unit: string;
  sourceUnit: string;
  sheet: string;
  kind: string;
  monthly: boolean;
  houseFilter: boolean;
  records: RecordValue[];
  missing: number;
  duplicates: number;
}

export interface Payload {
  indicators: ParsedIndicator[];
  warnings: string[];
  summary: {
    indicators: number;
    records: number;
    missing: number;
    years: number[];
    houses: string[];
  };
}

type Cell = string | number | boolean | Date | null;

/** Minúsculas, espaços colapsados e sem acentos — como norm() no Python. */
export function norm(value: unknown): string {
  const text = String(value || "")
    .split(/\s+/)
    .filter(Boolean)
    .join(" ")
    .toLowerCase();
  return text.normalize("NFD").replace(/\p{Mn}/gu, "");
}

/** Converte para número; ausência vira null; qualquer outra coisa é recusa. */
export function number(value: Cell): number | null {
  if (value === null || value === undefined) return null;
  if (typeof value === "boolean") throw new WorkbookError("Valor booleano em coluna numérica.");
  if (value instanceof Date) throw new WorkbookError("Valor não numérico ou fórmula sem resultado válido.");
  let candidate: string | number = value;
  if (typeof candidate === "string") {
    const trimmed = candidate.trim();
    if (["", "-", "—", "–"].includes(trimmed)) return null;
    candidate = trimmed.replace(/ /g, "");
    if (candidate.includes(",")) candidate = candidate.replace(/\./g, "").replace(/,/g, ".");
    if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(candidate)) {
      throw new WorkbookError("Valor não numérico ou fórmula sem resultado válido.");
    }
  }
  const result = Number(candidate);
  if (Number.isNaN(result)) throw new WorkbookError("Valor não numérico ou fórmula sem resultado válido.");
  if (!Number.isFinite(result) || result < 0) {
    throw new WorkbookError("Valor negativo ou não finito em indicador não negativo.");
  }
  return result;
}

/** Barreiras do arquivo antes de abrir a planilha: tamanho, macros, vínculos, XML. */
export function validateArchive(raw: Uint8Array): void {
  let entries: Record<string, Uint8Array>;
  try {
    entries = unzipSync(raw);
  } catch {
    throw new WorkbookError("Excel inválido ou corrompido.");
  }
  const names = Object.keys(entries);
  const total = names.reduce((sum, name) => sum + entries[name].length, 0);
  if (names.length > 4000 || total > 100 * 1024 * 1024) {
    throw new WorkbookError("Excel excede o limite de conteúdo descompactado (100 MB / 4.000 arquivos).");
  }
  if (!names.includes("[Content_Types].xml") || !names.includes("xl/workbook.xml")) {
    throw new WorkbookError("O arquivo não é uma planilha XLSX válida.");
  }
  if (names.some((name) => name.toLowerCase().includes("vbaproject") || name.startsWith("xl/externalLinks/"))) {
    throw new WorkbookError("Macros e vínculos externos não são aceitos.");
  }
  const decoder = new TextDecoder("utf-8", { fatal: false });
  for (const name of names) {
    if (name.endsWith(".xml") || name.endsWith(".rels")) {
      const content = decoder.decode(entries[name]);
      if (content.includes("<!DOCTYPE") || content.includes("<!ENTITY")) {
        throw new WorkbookError("Estrutura XML não permitida.");
      }
    }
  }
}

function cellValue(sheet: XLSX.WorkSheet, row: number, col: number): Cell {
  const cell = sheet[XLSX.utils.encode_cell({ r: row, c: col })] as XLSX.CellObject | undefined;
  if (!cell || cell.v === undefined || cell.v === null) return null;
  // Fórmula sem resultado em cache e célula de erro chegam como texto, e são recusadas adiante.
  if (cell.t === "e") return String(cell.w ?? "#ERRO");
  if (cell.t === "d") return cell.v instanceof Date ? cell.v : new Date(String(cell.v));
  if (cell.t === "b") return Boolean(cell.v);
  if (cell.t === "n") return Number(cell.v);
  return String(cell.v);
}

function readBlock(sheet: XLSX.WorkSheet, maxRow: number): Cell[][] {
  const rows: Cell[][] = [];
  for (let r = 0; r < maxRow; r += 1) {
    const row: Cell[] = [];
    for (let c = 0; c < MAX_COLS; c += 1) row.push(cellValue(sheet, r, c));
    rows.push(row);
  }
  return rows;
}

function findTable(
  rows: Cell[][],
  spec: Indicator,
): { header: number; dims: Record<string, number>; columns: number[] } | null {
  for (let index = 0; index < rows.length; index += 1) {
    const headers = rows[index].map((v) => norm(v));
    if (!headers.includes("ano")) continue;
    const columns = headers.map((h, i) => (spec.value.test(h) ? i : -1)).filter((i) => i >= 0);
    if (!columns.length) continue;
    const dims: Record<string, number> = {};
    for (const name of ["ano", "mes", "casa", "unidade"]) {
      const at = headers.indexOf(name);
      if (at >= 0) dims[name] = at;
    }
    if (spec.kind !== "ratio" && !spec.code.startsWith("G.") && !("casa" in dims)) continue;
    if (["A.1.1.1", "A.2.1.1", "A.2.2.1"].includes(spec.code) && !("mes" in dims)) continue;
    if (spec.category) {
      const cats = headers.map((h, i) => (spec.category!.test(h) ? i : -1)).filter((i) => i >= 0);
      if (cats.length !== 1) continue;
      dims.category = cats[0];
    }
    if (spec.kind !== "wide" && columns.length !== 1) {
      throw new WorkbookError(`${spec.code}: cabeçalhos de valores ambíguos.`);
    }
    return { header: index, dims, columns };
  }
  return null;
}

export function parseWorkbook(raw: Uint8Array): Payload {
  validateArchive(raw);
  let workbook: XLSX.WorkBook;
  try {
    workbook = XLSX.read(raw, { type: "array", cellDates: true, cellFormula: false, cellHTML: false });
  } catch {
    throw new WorkbookError("Não foi possível ler o Excel.");
  }
  if (workbook.SheetNames.length > 150) throw new WorkbookError("Limite de 150 abas excedido.");

  const catalogCodes = new Set(CATALOG.map((s) => s.code));
  const sheets: { name: string; sheet: XLSX.WorkSheet; preview: Cell[][]; codes: Set<string>; maxRow: number }[] = [];
  for (const name of workbook.SheetNames) {
    const sheet = workbook.Sheets[name];
    if (!sheet || !sheet["!ref"]) continue;
    const range = XLSX.utils.decode_range(sheet["!ref"]);
    const maxRow = range.e.r + 1;
    const preview = readBlock(sheet, Math.min(100, maxRow));
    const codes = new Set<string>();
    for (const row of preview) {
      for (const value of row) {
        if (typeof value === "string" && /^[ASG]\.\d+\.\d+\.\d+$/.test(value.trim())) codes.add(value.trim());
      }
    }
    if ([...codes].some((code) => catalogCodes.has(code))) sheets.push({ name, sheet, preview, codes, maxRow });
  }

  const indicators: ParsedIndicator[] = [];
  const warnings: string[] = [];
  for (const spec of CATALOG) {
    const candidates = sheets
      .filter((entry) => entry.codes.has(spec.code))
      .map((entry) => ({ entry, table: findTable(entry.preview, spec) }))
      .filter((c): c is { entry: (typeof sheets)[number]; table: NonNullable<ReturnType<typeof findTable>> } =>
        Boolean(c.table),
      );
    if (candidates.length !== 1) {
      throw new WorkbookError(
        `${spec.code}: indicador ausente, estrutura incompatível ou tabelas ambíguas (${candidates.length} correspondências).`,
      );
    }
    const { entry, table } = candidates[0];
    const { header, dims, columns } = table;
    if (entry.maxRow > MAX_ROWS) throw new WorkbookError(`${spec.code}: limite de ${MAX_ROWS} linhas excedido.`);
    const meta = entry.preview.find((row) => row.some((v) => String(v).trim() === spec.code))!;
    const codeCol = meta.findIndex((v) => String(v).trim() === spec.code);
    const unit = String(meta[codeCol + 3] || "");
    if (!unit) throw new WorkbookError(`${spec.code}: unidade de medida não identificada.`);
    // A unidade da tabela é mais específica que a dos metadados no caso do GEE (tCO2e).
    const headerValue = String(entry.preview[header][columns[0]]);
    const match = headerValue.match(/\(([^()]+)\)\s*$/);
    const displayUnit = match ? match[1] : unit;

    const records: RecordValue[] = [];
    const duplicates = new Map<string, number>();
    let missing = 0;
    const hasMonth = "mes" in dims;
    const relevant = new Set([...Object.values(dims), ...columns]);
    const lastRow = Math.min(entry.maxRow || MAX_ROWS, MAX_ROWS) + 1;
    for (let rowNo = header + 2; rowNo <= lastRow; rowNo += 1) {
      const row: Cell[] = [];
      for (let c = 0; c < MAX_COLS; c += 1) row.push(cellValue(entry.sheet, rowNo - 1, c));
      if (![...relevant].some((i) => row[i] !== null)) continue;
      if (rowNo > MAX_ROWS) throw new WorkbookError(`${spec.code}: limite de ${MAX_ROWS} linhas excedido.`);
      try {
        const yr = number(row[dims.ano]);
        if (yr === null || yr !== Math.trunc(yr) || yr < 1900 || yr > 2200) {
          throw new WorkbookError("Ano ausente ou inválido.");
        }
        const year = Math.trunc(yr);
        let month: number | null = null;
        if (hasMonth) {
          const text = norm(row[dims.mes]);
          if (!MONTHS.includes(text)) throw new WorkbookError("Mês ausente ou inválido.");
          month = MONTHS.indexOf(text) + 1;
        }
        const house = "casa" in dims && row[dims.casa] !== null ? String(row[dims.casa]).trim() : null;
        if ("casa" in dims && !house) throw new WorkbookError("Casa ausente.");
        const location =
          "unidade" in dims && row[dims.unidade] !== null ? String(row[dims.unidade]).trim() : null;
        for (const col of columns) {
          const value = number(row[col]);
          let category: string | null = null;
          if (spec.kind === "wide") {
            category = String(entry.preview[header][col])
              .replace(/^Número de colaboradores\s*/i, "")
              .replace(/^(do gênero|de gênero)\s+/i, "");
          } else if ("category" in dims) {
            if (row[dims.category] === null) throw new WorkbookError("Categoria ausente.");
            category = String(row[dims.category]).trim();
            if (spec.code === "A.5.1.1") category = `Escopo ${category}`;
          }
          if (value === null) missing += 1;
          const key = JSON.stringify([year, month, house, location, category, value]);
          duplicates.set(key, (duplicates.get(key) ?? 0) + 1);
          records.push({
            year,
            month,
            house,
            location,
            category,
            value,
            cell: `${XLSX.utils.encode_col(col)}${rowNo}`,
          });
        }
      } catch (error) {
        if (error instanceof WorkbookError) {
          throw new WorkbookError(`${spec.code}, linha ${rowNo}: ${error.message}`);
        }
        throw error;
      }
    }
    let repeated = 0;
    for (const count of duplicates.values()) if (count > 1) repeated += count - 1;
    if (missing) warnings.push(`${spec.code}: ${missing} valores ausentes; não convertidos em zero.`);
    if (repeated) warnings.push(`${spec.code}: ${repeated} registros iguais preservados; revisar na fonte.`);
    if (!records.length) warnings.push(`${spec.code}: estrutura válida, sem registros.`);
    if (hasMonth) {
      const periods = new Map<string, Set<number>>();
      for (const record of records) {
        const key = JSON.stringify([record.year, record.house]);
        if (!periods.has(key)) periods.set(key, new Set());
        periods.get(key)!.add(record.month as number);
      }
      let incomplete = 0;
      for (const months of periods.values()) if (months.size < 12) incomplete += 1;
      if (incomplete) warnings.push(`${spec.code}: ${incomplete} combinações Casa/ano com menos de 12 meses.`);
    }
    if (spec.kind === "ratio") {
      warnings.push(`${spec.code}: valor anual consolidado, sem recorte por Casa; proporções não são somadas.`);
    }
    indicators.push({
      code: spec.code,
      title: spec.title,
      unit: displayUnit,
      sourceUnit: unit,
      sheet: entry.name,
      kind: spec.kind,
      monthly: hasMonth,
      houseFilter: "casa" in dims,
      records,
      missing,
      duplicates: repeated,
    });
  }

  const all = indicators.flatMap((i) => i.records);
  return {
    indicators,
    warnings,
    summary: {
      indicators: indicators.length,
      records: all.length,
      missing: all.filter((r) => r.value === null).length,
      years: [...new Set(all.map((r) => r.year))].sort((a, b) => a - b),
      houses: [...new Set(all.map((r) => r.house).filter((h): h is string => Boolean(h)))].sort(),
    },
  };
}

File: backend/app/parser.py
"""Bounded, read-only XLSX extraction; cached values only, with cell provenance."""

import math
import re
import unicodedata
import zipfile
from collections import Counter
from pathlib import Path
from typing import Any

import openpyxl
from openpyxl.utils import get_column_letter

from .catalog import CATALOG, Indicator

MAX_ROWS = 30000
MAX_COLS = 64
MONTHS = [
    "janeiro",
    "fevereiro",
    "marco",
    "abril",
    "maio",
    "junho",
    "julho",
    "agosto",
    "setembro",
    "outubro",
    "novembro",
    "dezembro",
]


class WorkbookError(ValueError):
    pass


def norm(value: Any) -> str:
    text = " ".join(str(value or "").split()).lower()
    return "".join(c for c in unicodedata.normalize("NFD", text) if unicodedata.category(c) != "Mn")


def number(value: Any) -> float | None:
    if value is None or (isinstance(value, str) and value.strip() in ("", "-", "—", "–")):
        return None
    if isinstance(value, bool):
        raise WorkbookError("Valor booleano em coluna numérica.")
    if isinstance(value, str):
        value = value.strip().replace("\u00a0", "")
        if "," in value:
            value = value.replace(".", "").replace(",", ".")
    try:
        result = float(value)
    except (TypeError, ValueError):
        raise WorkbookError("Valor não numérico ou fórmula sem resultado válido.") from None
    if not math.isfinite(result) or result < 0:
        raise WorkbookError("Valor negativo ou não finito em indicador não negativo.")
    return result


def validate_archive(path: Path) -> None:
    try:
        with zipfile.ZipFile(path) as archive:
            entries = archive.infolist()
            if len(entries) > 4000 or sum(e.file_size for e in entries) > 100 * 1024 * 1024:
                raise WorkbookError("Excel excede o limite de conteúdo descompactado (100 MB / 4.000 arquivos).")
            names = {e.filename for e in entries}
            if not {"[Content_Types].xml", "xl/workbook.xml"}.issubset(names):
                raise WorkbookError("O arquivo não é uma planilha XLSX válida.")
            if any("vbaproject" in n.lower() or n.startswith("xl/externalLinks/") for n in names):
                raise WorkbookError("Macros e vínculos externos não são aceitos.")
            for entry in entries:
                if entry.flag_bits & 1:
                    raise WorkbookError("Planilhas criptografadas não são aceitas.")
                if entry.filename.endswith((".xml", ".rels")):
                    content = archive.read(entry)
                    if b"<!DOCTYPE" in content or b"<!ENTITY" in content:
                        raise WorkbookError("Estrutura XML não permitida.")
            if archive.testzip():
                raise WorkbookError("Excel corrompido.")
    except (zipfile.BadZipFile, OSError, RuntimeError):
        raise WorkbookError("Excel inválido ou corrompido.") from None


def find_table(rows: list[tuple[Any, ...]], spec: Indicator) -> tuple[int, dict[str, int], list[int]] | None:
    for index, row in enumerate(rows):
        headers = [norm(v) for v in row]
        if "ano" not in headers:
            continue
        values = [i for i, h in enumerate(headers) if re.search(spec.value, h)]
        if not values:
            continue
        dims: dict[str, int] = {h: headers.index(h) for h in ("ano", "mes", "casa", "unidade") if h in headers}
        if spec.kind != "ratio" and not spec.code.startswith("G.") and "casa" not in dims:
            continue
        if spec.code in ("A.1.1.1", "A.2.1.1", "A.2.2.1") and "mes" not in dims:
            continue
        if spec.category:
            cats = [i for i, h in enumerate(headers) if re.search(spec.category, h)]
            if len(cats) != 1:
                continue
            dims["category"] = cats[0]
        if spec.kind != "wide" and len(values) != 1:
            raise WorkbookError(f"{spec.code}: cabeçalhos de valores ambíguos.")
        return index, dims, values
    return None


def parse_workbook(path: Path) -> dict[str, Any]:
    validate_archive(path)
    try:
        workbook = openpyxl.load_workbook(path, read_only=True, data_only=True, keep_links=False)
    except Exception:
        raise WorkbookError("Não foi possível ler o Excel.") from None
    try:
        if len(workbook.sheetnames) > 150:
            raise WorkbookError("Limite de 150 abas excedido.")
        # Formatting in the supplied workbook extends to XFD. Bound reads to the
        # documented supported data area, independently of misleading dimensions.
        sheets = []
        for sheet in workbook:
            preview = list(sheet.iter_rows(max_row=100, max_col=MAX_COLS, values_only=True))
            codes = {
                str(v).strip()
                for row in preview
                for v in row
                if isinstance(v, str) and re.fullmatch(r"[ASG]\.\d+\.\d+\.\d+", v.strip())
            }
            if codes.intersection(s.code for s in CATALOG):
                sheets.append((sheet, preview, codes))
        indicators: list[dict[str, Any]] = []
        warnings = []
        for spec in CATALOG:
            candidates = []
            for sheet, preview, codes in sheets:
                if spec.code in codes:
                    table = find_table(preview, spec)
                    if table:
                        candidates.append((sheet, preview, table))
            if len(candidates) != 1:
                raise WorkbookError(
                    f"{spec.code}: indicador ausente, estrutura incompatível ou tabelas ambíguas ({len(candidates)} correspondências)."
                )
            sheet, preview, (header, dims, columns) = candidates[0]
            if (sheet.max_row or 0) > MAX_ROWS:
                raise WorkbookError(f"{spec.code}: limite de {MAX_ROWS} linhas excedido.")
            meta = next(row for row in preview if any(str(v).strip() == spec.code for v in row))
            code_col = next(i for i, v in enumerate(meta) if str(v).strip() == spec.code)
            unit = str(meta[code_col + 3] or "")
            if not unit:
                raise WorkbookError(f"{spec.code}: unidade de medida não identificada.")
            # Table units are more specific than metadata for GEE (tCO2e).
            header_value = str(preview[header][columns[0]])
            match = re.search(r"\(([^()]+)\)\s*$", header_value)
            display_unit = match.group(1) if match else unit
            records: list[dict[str, Any]] = []
            duplicates: Counter[tuple[Any, ...]] = Counter()
            missing = 0
            has_month = "mes" in dims
            relevant = set(dims.values()) | set(columns)
            for row_no, row in enumerate(
                sheet.iter_rows(
                    min_row=header + 2,
                    max_row=min(sheet.max_row or MAX_ROWS, MAX_ROWS) + 1,
                    max_col=MAX_COLS,
                    values_only=True,
                ),
                header + 2,
            ):
                if not any(row[i] is not None for i in relevant):
                    continue
                if row_no > MAX_ROWS:
                    raise WorkbookError(f"{spec.code}: limite de {MAX_ROWS} linhas excedido.")
                try:
                    yr = number(row[dims["ano"]])
                    if yr is None or yr != int(yr) or not 1900 <= yr <= 2200:
                        raise WorkbookError("Ano ausente ou inválido.")
                    year = int(yr)
                    month = None
                    if has_month:
                        text = norm(row[dims["mes"]])
                        if text not in MONTHS:
                            raise WorkbookError("Mês ausente ou inválido.")
                        month = MONTHS.index(text) + 1
                    house = str(row[dims["casa"]]).strip() if "casa" in dims and row[dims["casa"]] is not None else None
                    if "casa" in dims and not house:
                        raise WorkbookError("Casa ausente.")
                    location = (
                        str(row[dims["unidade"]]).strip()
                        if "unidade" in dims and row[dims["unidade"]] is not None
                        else None
                    )
                    for col in columns:
                        value = number(row[col])
                        category = None
                        if spec.kind == "wide":
                            category = re.sub(r"^Número de colaboradores\s*", "", str(preview[header][col]), flags=re.I)
                            category = re.sub(r"^(do gênero|de gênero)\s+", "", category, flags=re.I)
                        elif "category" in dims:
                            if row[dims["category"]] is None:
                                raise WorkbookError("Categoria ausente.")
                            category = str(row[dims["category"]]).strip()
                            if spec.code == "A.5.1.1":
                                category = f"Escopo {category}"
                        missing += value is None
                        key = (year, month, house, location, category, value)
                        duplicates[key] += 1
                        records.append(
                            {
                                "year": year,
                                "month": month,
                                "house": house,
                                "location": location,
                                "category": category,
                                "value": value,
                                "cell": f"{get_column_letter(col + 1)}{row_no}",
                            }
                        )
                except WorkbookError as error:
                    raise WorkbookError(f"{spec.code}, linha {row_no}: {error}") from None
            repeated = sum(v - 1 for v in duplicates.values() if v > 1)
            if missing:
                warnings.append(f"{spec.code}: {missing} valores ausentes; não convertidos em zero.")
            if repeated:
                warnings.append(f"{spec.code}: {repeated} registros iguais preservados; revisar na fonte.")
            if not records:
                warnings.append(f"{spec.code}: estrutura válida, sem registros.")
            if has_month:
                periods: dict[tuple[Any, ...], set[int]] = {}
                for record in records:
                    periods.setdefault((record["year"], record["house"]), set()).add(record["month"])
                incomplete = sum(len(m) < 12 for m in periods.values())
                if incomplete:
                    warnings.append(f"{spec.code}: {incomplete} combinações Casa/ano com menos de 12 meses.")
            if spec.kind == "ratio":
                warnings.append(
                    f"{spec.code}: valor anual consolidado, sem recorte por Casa; proporções não são somadas."
                )
            indicators.append(
                {
                    "code": spec.code,
                    "title": spec.title,
                    "unit": display_unit,
                    "sourceUnit": unit,
                    "sheet": sheet.title,
                    "kind": spec.kind,
                    "monthly": has_month,
                    "houseFilter": "casa" in dims,
                    "records": records,
                    "missing": missing,
                    "duplicates": repeated,
                }
            )
        all_records = [r for i in indicators for r in i["records"]]
        return {
            "indicators": indicators,
            "warnings": warnings,
            "summary": {
                "indicators": len(indicators),
                "records": len(all_records),
                "missing": sum(r["value"] is None for r in all_records),
                "years": sorted({r["year"] for r in all_records}),
                "houses": sorted({r["house"] for r in all_records if r["house"]}),
            },
        }
    finally:
        workbook.close()
